### DISCUSSION TITLE
Register and login user without a password (custom oauth integration)

### DISCUSSION THREAD

**Discussion by saeedesmaili** [fastapi-users/fastapi-users#1157] [Q&A]
I read [this response](https://github.com/fastapi-users/fastapi-users/discussions/992#discussioncomment-2743311) and tried to utilize it in my attempt to implement the [google one tap](https://piraces.dev/posts/how-to-use-google-one-tap/) authentication.
The way explained in the mentioned response works when I create a custom registration route and ask for email and password, but not with my current workflow:
- User clicks on the one-tap pop-up and authenticates with google.
- Google returns an `id_token`.
- I decode the `id_token` using [this endpoint](https://oauth2.googleapis.com/tokeninfo?id_token={id_token}) (couldn't find a better way yet).
- It returns the user's info:
```
{
  "iss": "https://accounts.google.com",
  "nbf": "167...",
  "aud": "519...53-v0sd...929sqj.apps.googleusercontent.com",
  "sub": "1140....9029",
  "email": "__user's_email__",
  "email_verified": "true",
  "azp": "519456...9sqj.apps.googleusercontent.com",
  "name": "__user's_name__",
  "picture": "https://lh3.googleusercontent.com/a/AEdFTp6...xpA=s96-c",
  "given_name": "...",
  "family_name": "...",
  "iat": "16...380",
  "exp": "16...980",
  "jti": "68ab..b5432c38",
  "alg": "R....6",
  "kid": "274..1d90cfb585a",
  "typ": "JWT"
}
```
- Now I have the user's email and I can check if the user exists or not.
- If the user doesn't exist, I can register the user with an arbitrary password and log in with the method that I mentioned.
- But if the user already exists, I don't have the user's password in this step and can't log in them. The workflow stops here.

Is there a way for me to proceed with implementing this one-tap login?

**Comment by frankie567** [comment]
Why not using our OAuth integration instead? https://fastapi-users.github.io/fastapi-users/10.3/configuration/oauth/ We already handle this kind of things.

**Reply by saeedesmaili** [reply]
Yeah, I'm currently using that in my normal Google authentication (on the login page, when the user clicks on the Google button and I can send them to the authorization path generated by `/authorize`).
But I couldn't figure out how to do this with Google one-tap. When the user clicks on one of their Google accounts on the one-tap pop-up:
![image](https://user-images.githubusercontent.com/13351748/216787895-45da1d05-ed6d-4e33-95a8-aa428a69b4d0.png)
 I get the following data from Google:
<img width="168" alt="image" src="https://user-images.githubusercontent.com/13351748/216787943-0dbab93f-1ed7-40b0-9cf9-ee6957c7a33e.png">
And when I pass the `credential` as `id_token` as described in the original post, I get the response that I posted in the original post. So I don't have any of `code`, `code_verifier`, `state`, and `error` required by the callback route of fastapi-users.

**Reply by saeedesmaili** [reply]
More info about google's one-tap can be found on their doc: [Verify the Google ID token on your server side](https://developers.google.com/identity/gsi/web/guides/verify-google-id-token)

**Reply by frankie567** [reply]
I see. Well, it's a completely different flow from the one we provide here. You should probably implement your own callback to handle the tokens from Google. What you could to circumvent the lack of password is to generate a random one. This is what we do for OAuth2 callback: https://github.com/fastapi-users/fastapi-users/blob/024e6cd4516a55c2205fc58f719db1e581dd8078/fastapi_users/manager.py#L204-L213

**Reply by saeedesmaili** [reply]
Yeah, and I've tried to do exactly this (though didn't know about this code snippet and was generating my own random password).
The main question here is how can I then implement something like what [you've explained in response to this question](https://github.com/fastapi-users/fastapi-users/discussions/992#discussioncomment-2743311) (to generate access token)?
1. If it's a new user, I have just created the random password so I can potentially pass the password to the following code somehow? (but how?)
```python
created_user = await user_manager.create(
    user_create, safe=True, request=request
)
```
2. If the user already exists and the account has been created with this method, I must store and find the randomly generated password for the user I guess (it doesn't seem like a best practice, right?)
3. If the user already exists but the account has been created with a normal registration (the user has entered their own password) or they have changed their password, I'm completely screwed in this case and can't log in them using the above code.

I know this seems like an edge case, but I'm asking here to know what you think the best approach for this could be. I might just get rid of adding the google one-tap at the end (and hope that it gets supported by fastapi-users or httpx-oauth some day).
